druid-shell/src/windows/application.rs

```diff
@@ -14,10 +14,121 @@
 
 //! Windows implementation of features at the application scope.
 
+use std::ptr;
+
+use winapi::shared::minwindef::FALSE;
+use winapi::shared::minwindef::UINT;
+use winapi::shared::ntdef::{LPWSTR, WCHAR};
+use winapi::shared::winerror::ERROR_SUCCESS;
+use winapi::um::errhandlingapi::*;
+use winapi::um::winbase::{GlobalAlloc, GlobalLock, GlobalUnlock, GMEM_MOVEABLE};
+use winapi::um::winuser::*;
+
+use crate::clipboard::ClipboardItem;
+use crate::util::{FromWide, ToWide};
+
 pub struct Application;
 
 impl Application {
     pub fn quit() {
         crate::runloop::request_quit();
     }
+
+    /// Returns the contents of the clipboard, if any.
+    pub fn get_clipboard_contents() -> Option<ClipboardItem> {
+        unsafe {
+            if OpenClipboard(ptr::null_mut()) == FALSE {
+                return None;
+            }
+
+            let result = get_clipboard_impl();
+            CloseClipboard();
+            result
+        }
+    }
+
+    pub fn set_clipboard_contents(item: ClipboardItem) {
+        unsafe {
+            if OpenClipboard(ptr::null_mut()) == FALSE {
+                return;
+            }
+            EmptyClipboard();
+            match item {
+                ClipboardItem::Text(string) => {
+                    let wstr = string.to_wide();
```

Comment:
This is ok but as I've mentioned elsewhere I'd like to migrate from our util wide methods to those in `wio`.

Reply:
right, I was definitely fumbling in the dark here. Also not totally confident about coping the data between the pointers, although it seems okay?

Reply:
The copying looks fine to my eyes. You could get away with one fewer copy but I think it's not worth it; better to keep the unsafe block small.